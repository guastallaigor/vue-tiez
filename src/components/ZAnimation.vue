<template>
  <div :class="{ dark }" ref="animation" class="z-animation">
    <svg
      :style="{ top: y - 10, left: x - 10 }"
      width="20"
      height="20"
      viewBox="0 0 20 20"
      version="1.1"
      xmlns="http://www.w3.org/2000/svg"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    >
      <g stroke="none" strokeWidth="1" fill="none" fillRule="evenodd">
        <g :fill="dark ? '#333' : '#ddd'">
          <rect width="100%" height="100%" rx="10" />
        </g>
      </g>
    </svg>
  </div>
</template>

<script>
export default {
  name: 'z-animation',
  props: {
    dark: {
      type: Boolean,
      default: false
    },
    y: {
      type: [String, Number],
      default: 10
    },
    x: {
      type: [String, Number],
      default: 10
    }
  },
  mounted () {
    this.addAnimationEvent()
  },
  beforeDestroy () {
    this.removeAnimationEvent()
  },
  methods: {
    addAnimationEvent () {
      if (this.$refs.animation) {
        this.$refs.animation.addEventListener('animationend', this.onAnimationEnd)
      }
    },
    removeAnimationEvent () {
      if (this.$refs.animation) {
        this.$refs.animation.removeEventListener('animationend', this.onAnimationEnd)
      }
    },
    onAnimationEnd (event) {
      this.$emit('complete', event || null)
    }
  }
}
</script>

<style lang="scss">
.z-animation {
  @keyframes expand {
    0% {
      opacity: 0;
      transform: scale(1);
    }

    30% {
      opacity: 1;
    }

    80% {
      opacity: 0.5;
    }

    100% {
      transform: scale(25);
      opacity: 0;
    }
  }

  div {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
  }

  svg {
    position: absolute;
    animation: 400ms ease-in expand;
    animation-fill-mode: forwards;
    width: 20px;
    height: 20px;
  }
}
</style>
