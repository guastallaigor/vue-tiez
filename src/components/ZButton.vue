<template>
  <button
    role="button"
    type="button"
    class="z-button"
    @click="onClick"
    :class="getClasses"
    ref="button"
    @mouseover="hover = true"
    @mouseleave="hover = false"
  >
    <span v-if="loading" class="loading-shim">
      <span class="loading">
        <span></span>
        <span></span>
        <span></span>
      </span>
    </span>
    <span v-if="icon && !loading && !withIconRight" class="icon" :class="{'right': withIconRight}">
      <z-icon :name="icon" :color="getIconColor" />
    </span>
    <span v-if="content && !loading">{{ content }}</span>
    <slot v-if="!content && !loading" />
    <span v-if="icon && !loading && withIconRight" class="icon" :class="{'right': withIconRight}">
      <z-icon :name="icon" :color="getIconColor" />
    </span>
    <z-animation v-if="x && y && !shadow" :dark="dark" :y="y" :x="x" @complete="onAnimationComplete" />
  </button>
</template>

<script>
export default {
  name: 'z-button',
  components: {
    ZIcon: () => import('./ZIcon'),
    ZAnimation: () => import('./ZAnimation')
  },
  props: {
    content: {
      type: String,
      default: ''
    },
    icon: {
      type: String,
      default: ''
    },
    withIconRight: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    small: {
      type: Boolean,
      default: false
    },
    loading: {
      type: Boolean,
      default: false
    },
    active: {
      type: Boolean,
      default: false
    },
    dark: {
      type: Boolean,
      default: false
    },
    shadow: {
      type: Boolean,
      default: false
    },
    type: {
      type: String,
      default: 'primary',
      validator: (v) => (['primary', 'secondary', 'warning', 'highlight', 'link', 'abort'].includes(v))
    }
  },
  data: () => ({
    x: null,
    y: null,
    hover: false
  }),
  computed: {
    getClasses () {
      const { icon, disabled, dark, type, small, loading, shadow, active } = this

      return {
        icon,
        disabled,
        dark,
        small,
        loading,
        shadow,
        active,
        primary: type === 'primary',
        secondary: type === 'secondary',
        warning: type === 'warning',
        highlight: type === 'highlight',
        link: type === 'link',
        abort: type === 'abort'
      }
    },
    getIconColor () {
      const { type, hover } = this

      return {
        primary: hover ? '#000' : '#fff',
        secondary: hover ? '#000' : 'rgb(102, 102, 102)',
        warning: hover ? '#eb5757' : '#fff',
        highlight: hover ? '#007aff' : '#fff',
        link: '#0d7df7',
        abort: '#666'
      }[type]
    }
  },
  methods: {
    onClick (event) {
      if (this.disabled) return

      const rect = this.$refs.button.getBoundingClientRect()

      if (rect) {
        const { left, top } = rect
        this.x = event.clientX - left
        this.y = event.clientY - top
      }

      this.$emit('click', event)
    },
    onAnimationComplete () {
      this.x = null
      this.y = null
    }
  }
}
</script>

<style lang="scss">
@mixin set-color-bg($color, $bg, $isbg: true) {
  color: $color;

  @if ($isbg == true) {
    background: $bg;
  }

  @else {
    background-color: $bg;
  }
}

@mixin set-border-color-bg($border: rgb(0, 0, 0), $color: rgb(0, 0, 0), $bg: rgb(255, 255, 255), $isbg: true) {
  border: 1px solid $border;

  @include set-color-bg($color, $bg, $isbg);
}

@mixin set-bordercolor-color-bg($bordercolor: rgb(0, 0, 0), $color: rgb(0, 0, 0), $bg: rgb(255, 255, 255), $isbg: true) {
  border-color: $bordercolor;

  @include set-color-bg($color, $bg, $isbg);
}

@mixin set-link($color: #0d7df7) {
  border: 0;
  background: transparent;
  color: $color;
  padding: 0;
  box-shadow: none;
}

.z-button {
  @keyframes hasloading {
    0% {
      opacity: 0.2;
    }

    20% {
      opacity: 1;
    }

    100% {
      opacity: 0.2;
    }
  }

  appearance: none;
  align-items: center;
  color: rgb(255, 255, 255);
  background: rgb(0, 0, 0);
  display: inline-flex;
  min-width: 200px;
  height: 40px;
  padding: 0 25px;
  outline: none;
  border: 1px solid rgb(0, 0, 0);
  font-size: 12px;
  justify-content: center;
  text-transform: uppercase;
  cursor: pointer;
  text-align: center;
  user-select: none;
  font-weight: 100;
  position: relative;
  overflow: hidden;
  transition: border 0.2s, background 0.2s, color 0.2s ease-out;
  border-radius: 5px;
  white-space: nowrap;
  text-decoration: none;
  line-height: 0;

  span {
    position: relative;
    z-index: 1;
  }

  &.secondary {
    @include set-border-color-bg(#eaeaea, rgb(102, 102, 102), #fff);

    border-color: rgb(234, 234, 234);

    &.dark {
      @include set-border-color-bg(#666, rgb(234, 234, 234), #000);

      &:hover:not(.shadow) {
        @include set-border-color-bg(#fff, rgb(0, 0, 0), #fff);
      }
    }

    &:hover:not(.shadow),
    &.active:not(.shadow) {
      @include set-bordercolor-color-bg(rgb(102, 102, 102), rgb(102, 102, 102), #fff);
    }

    &.disabled {
      &:hover {
        @include set-bordercolor-color-bg(#eaeaea, #ccc, #fafafa, false);
      }
    }
  }

  .icon {
    position: absolute;
    left: 8px;
    right: auto;
    display: flex;
    align-items: center;
    top: 0;
    bottom: 0;
    z-index: 1;

    &.right {
      left: auto;
      right: 8px;
    }
  }

  b {
    display: inline-block;
    overflow: none;
    z-index: 100;
    font-weight: 500;
    position: relative;
  }

  &.dark {
    border: 2px solid #fff;

    @include set-color-bg(#000, #fff);

    &.disabled,
    &.disabled:hover:not(.shadow) {
      @include set-border-color-bg(#333, #333, #111);
    }

    &:hover:not(.shadow),
    &.active {
      @include set-border-color-bg(#fff, #fff, rgb(0, 0, 0));
    }

    &.loading {
      @include set-bordercolor-color-bg(rgb(234, 234, 234), rgb(204, 204, 204), rgb(250, 250, 250));
    }
  }

  &.loading {
    @include set-bordercolor-color-bg(#eaeaea, #ccc, #fafafa);

    pointer-events: none;
    position: relative;
    cursor: default;

    .loading-shim {
      position: absolute;
      top: 0;
      right: 0;
      left: 0;
      bottom: 0;
      display: flex;
      align-items: center;
      justify-content: center;

      .loading {
        span {
          animation-name: hasloading;
          animation-iteration-count: infinite;
          animation-fill-mode: both;
          width: 4px;
          height: 4px;
          border-radius: 50%;
          background-color: #444;
          display: inline-block;
          margin: 0 1px;
          animation-duration: 1.4s;

          &:nth-child(2) {
            animation-delay: 0.2s;
          }

          &:nth-child(3) {
            animation-delay: 0.4s;
          }
        }
      }
    }
  }

  &.abort {
    border: 0;

    @include set-bordercolor-color-bg(transparent, #666, transparent, false);

    &.disabled {
      color: #ccc;
    }

    &:hover:not(.shadow) {
      @include set-link(#666);
    }
  }

  &:hover:not(.shadow),
  &.active:not(.shadow) {
    @include set-border-color-bg();
  }

  &.warning {
    @include set-bordercolor-color-bg(#eb5757, #fff, #eb5757, false);

    &:hover:not(.shadow),
    &.active:not(.shadow) {
      @include set-border-color-bg(#eb5757, #eb5757, #fff);
    }
  }

  &.highlight {
    @include set-bordercolor-color-bg(#007aff, #fff, #007aff, false);

    &:hover:not(.shadow),
    &.active:not(.shadow) {
      @include set-border-color-bg(#007aff, #007aff, #fff);
    }
  }

  &.disabled {
    @include set-bordercolor-color-bg(#eaeaea, #ccc, #fafafa);

    &:hover:not(.shadow) {
      &:not(.link):not(.dark) {
        @include set-border-color-bg(#eaeaea, #ccc, #fafafa);
      }
    }

    cursor: not-allowed;
  }

  &.small {
    min-width: 100px;
    height: 24px;
    line-height: 22px;
    padding: 0 10px;
  }

  &.shadow {
    font-weight: 400;
    box-shadow: 0 5px 10px rgba(0, 0, 0, 0.12);

    &:not(.disabled) {
      &:hover {
        box-shadow: 0 7px 20px rgba(0, 0, 0, 0.16);
        transform: translate3d(0, -1px, 0);
      }
    }
  }

  &.link {
    @include set-link();

    &.disabled {
      color: #ccc;
    }

    &:hover:not(.shadow) {
      @include set-link();

      text-decoration: underline;

      &:not(.disabled) {
        color: #0d7df7;
      }

      &.disabled {
        color: #ccc;
      }
    }
  }
}
</style>
