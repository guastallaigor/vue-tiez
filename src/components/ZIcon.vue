<template>
  <component :is="name" :size="size" :dark="dark"/>
</template>

<script>
export default {
  name: 'z-icon',
  components: {
    add: () => import('./icons/add'),
    arrow: () => import('./icons/arrow'),
    'arrow-double': () => import('./icons/arrow-double')
  },
  props: {
    name: {
      type: String,
      required: true
    },
    size: {
      type: String,
      default: '20px'
    },
    dark: {
      type: Boolean,
      default: false
    }
  }
}
</script>

<style lang="scss">
svg.z-icon {
  shape-rendering: crispEdges;

  circle {
    transition: all 0.2s ease 0s;
  }

  &:hover circle {
    stroke: rgb(153, 153, 153);
  }
}
</style>
