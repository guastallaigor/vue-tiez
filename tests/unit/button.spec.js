import { shallowMount, mount } from '@vue/test-utils'
import ZButton from '../../src/components/ZButton'
import ZAnimation from '../../src/components/ZAnimation'
import ZIcon from '../../src/components/ZIcon'

describe('When I create the ZButton component', () => {
  const createComponent = (propsData = {}, slot) => {
    return shallowMount(ZButton, {
      propsData: propsData,
      stubs: { ZAnimation, ZIcon },
      slots: {
        default: slot || 'ZButton'
      }
    })
  }

  it('should be a Vue instance', () => {
    const wrapper = createComponent({ content: 'Test' })
    expect(wrapper.find(ZButton).isVueInstance()).toBe(true)
  })

  it('should have an icon', (done) => {
    const wrapper = mount(ZButton, {
      AsyncComponent: ZIcon,
      propsData: { content: 'test', icon: 'add' }
    })
    const button = wrapper.find('.z-button')
    expect(button.exists()).toBe(true)
    expect(button.classes().length).toBe(3)
    expect(button.classes()).toEqual(['z-button', 'icon', 'primary'])
    setTimeout(() => {
      expect(wrapper.find(ZIcon).exists()).toBe(true)
      const icon = wrapper.find('.z-button > .icon')
      expect(icon.exists()).toBe(true)
      const circle = wrapper.find('.z-button > .icon > svg > circle')
      expect(circle.attributes().stroke).toBe('#fff')
      // const line = wrapper.findAll('.z-input > .wrapper > icon > svg > line')
      // expect(line.at(0).attributes().stroke).toBe('#fff')
      // expect(line.at(1).attributes().stroke).toBe('#fff')
      done()
    })
  })

  // it('should have a dark class', () => {
  //   const wrapper = createComponent({ value: '', dark: true })
  //   const inputLabel = wrapper.find('.z-input > .wrapper')
  //   expect(inputLabel.exists()).toBe(true)
  //   expect(inputLabel.classes().length).toBe(2)
  //   expect(inputLabel.classes()).toEqual(['wrapper', 'dark'])
  // })

  // it('should have a dark class and have a label', () => {
  //   const wrapper = createComponent({ value: '', dark: true, label: 'Label' })
  //   const label = wrapper.find('.z-input > .label')
  //   expect(label.exists()).toBe(true)
  //   expect(label.classes().length).toBe(2)
  //   expect(label.classes()).toEqual(['label', 'dark'])
  // })

  // it('should have a label', () => {
  //   const wrapper = createComponent({ value: '', label: 'Label' })
  //   const label = wrapper.find('.z-input > .label')
  //   expect(label.exists()).toBe(true)
  //   expect(label.classes().length).toBe(1)
  //   expect(label.classes()).toEqual(['label'])
  //   const input = wrapper.find('.z-input')
  //   expect(input.exists()).toBe(true)
  //   expect(input.classes().length).toBe(1)
  //   expect(input.attributes().style).toBe('display: inline-flex;')
  //   const wrapperDiv = wrapper.find('.z-input > .wrapper')
  //   expect(wrapperDiv.exists()).toBe(true)
  //   expect(wrapperDiv.classes().length).toBe(2)
  //   expect(wrapperDiv.classes()).toEqual(['wrapper', 'inner-input'])
  // })

  // it('should have a type text', async () => {
  //   const wrapper = createComponent({ value: '', type: 'text' })
  //   const input = wrapper.find('.z-input > .wrapper > div > input')
  //   expect(input.exists()).toBe(true)
  //   expect(input.attributes().type).toBe('text')
  //   input.element.value = 'test'
  //   input.trigger('input')
  //   await wrapper.vm.$nextTick()
  //   expect(wrapper.vm.$refs.input.value).toBe('test')
  //   expect(wrapper.emitted().input).toBeTruthy()
  //   expect(wrapper.emitted().input[0]).toEqual(['test'])
  // })

  // it('should have a type number', async () => {
  //   const wrapper = createComponent({ value: '', type: 'number' })
  //   const input = wrapper.find('.z-input > .wrapper > div > input')
  //   expect(input.exists()).toBe(true)
  //   expect(input.attributes().type).toBe('number')
  //   input.element.value = 123
  //   input.trigger('input')
  //   await wrapper.vm.$nextTick()
  //   expect(wrapper.vm.$refs.input.value).toBe('123')
  //   expect(wrapper.emitted().input).toBeTruthy()
  //   expect(wrapper.emitted().input[0]).toEqual(['123'])
  // })

  // it('should have a maxLength of 3 characters', async () => {
  //   const wrapper = createComponent({ value: '', maxLength: 3 })
  //   const input = wrapper.find('.z-input > .wrapper > div > input')
  //   expect(input.exists()).toBe(true)
  //   expect(input.attributes().type).toBe('text')
  //   expect(input.attributes().maxlength).toBe('3')
  //   wrapper.vm.valueComp = 'teste'
  //   await wrapper.vm.$nextTick()
  //   expect(wrapper.emitted().input).toBeTruthy()
  //   expect(wrapper.emitted().input[0]).toEqual(['tes'])
  // })

  // it('should have a disable attribute and class', async () => {
  //   const wrapper = createComponent({ value: '', disabled: true })
  //   const input = wrapper.find('.z-input > .wrapper > div > input')
  //   expect(input.exists()).toBe(true)
  //   expect(input.attributes().type).toBe('text')
  //   expect(input.attributes().disabled).toBe('disabled')
  //   const wrapperDiv = wrapper.find('.z-input > .wrapper')
  //   expect(wrapperDiv.exists()).toBe(true)
  //   expect(wrapperDiv.classes().length).toBe(2)
  //   expect(wrapperDiv.classes()).toEqual(['wrapper', 'disabled'])
  //   input.element.value = 'test'
  //   input.trigger('input')
  //   await wrapper.vm.$nextTick()
  //   wrapper.vm.valueComp = 'other'
  //   expect(wrapper.vm.valueComp).toBe('')
  //   expect(wrapper.emitted().input).toBeFalsy()
  // })

  // it('should have a placeholder', () => {
  //   const wrapper = createComponent({ value: '', placeholder: 'Placeholder' })
  //   const input = wrapper.find('.z-input > .wrapper > div > input')
  //   expect(input.exists()).toBe(true)
  //   expect(input.attributes().placeholder).toBe('Placeholder')
  // })

  // it('should have an error class', () => {
  //   const wrapper = createComponent({ value: '', error: true })
  //   const wrapperDiv = wrapper.find('.z-input > .wrapper')
  //   expect(wrapperDiv.exists()).toBe(true)
  //   expect(wrapperDiv.classes().length).toBe(2)
  //   expect(wrapperDiv.classes()).toEqual(['wrapper', 'error'])
  // })

  // it('should emit a blur event', async () => {
  //   const wrapper = createComponent({ value: '' })
  //   const input = wrapper.find('.z-input > .wrapper > div > input')
  //   expect(input.exists()).toBe(true)
  //   expect(input.attributes().type).toBe('text')
  //   input.element.value = 'test'
  //   input.trigger('blur')
  //   await wrapper.vm.$nextTick()
  //   expect(wrapper.emitted().blur).toBeTruthy()
  //   expect(wrapper.vm.focused).toBe(false)
  // })

  // it('should emit a focus event', async () => {
  //   const wrapper = createComponent({ value: '', autoFocus: true })
  //   const input = wrapper.find('.z-input > .wrapper > div > input')
  //   expect(input.exists()).toBe(true)
  //   expect(input.attributes().type).toBe('text')
  //   input.trigger('focus')
  //   await wrapper.vm.$nextTick()
  //   expect(wrapper.emitted().focus).toBeTruthy()
  //   expect(wrapper.vm.focused).toBe(true)
  // })

  it('should match snapshot', () => {
    const wrapper = createComponent({
      content: 'Test',
      icon: '',
      withIconRight: false,
      disabled: false,
      small: false,
      loading: false,
      active: false,
      dark: false,
      shadow: false,
      type: 'primary'
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
