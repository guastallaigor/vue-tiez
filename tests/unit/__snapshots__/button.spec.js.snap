// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When I create the ZButton component should match snapshot 1`] = `
<button role="button" type="button" class="z-button primary">
  <!---->
  <!----> <span>Test</span>
  <!---->
  <!---->
  <!----></button>
`;
